test/realtime/message.test.js

```diff
@@ -1183,5 +1183,47 @@ define(['ably', 'shared_helper', 'async', 'chai'], function (Ably, helper, async
           });
       });
     }
+
+    it('subscribes to filtered channel', function (done) {
+      var testData = {
+        name: 'filtered',
+        data: 'This should be filtered',
+        extras: {
+            headers: {
+                name: "value1",
+                number: 26095,
+                bool: true,
+            }
+        }
+      }
+      var filterOption = {
+        filter: 'name == `"filtered"` && headers.number == `26095`'
+      }
+      try {
+        /* set up realtime */
+        var realtime = helper.AblyRealtime();
+
+        realtime.connection.on('connected', function () {
+          var rtChannel = realtime.channels.get('filter_test', filterOption);
+
+            /* subscribe to event */
+            rtChannel.subscribe('filtered', function (msg) {
+              try {
+                expect(msg.data).to.equal(testData.data, 'Unexpected msg text received');
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+              closeAndFinish(done, realtime);
+            });
+
+            rtChannel.publish("filtered", "no filter on message");
```

Comment:
Given you are publishing the non-filtered message first, `closeAndFinish(done, realtime)` i.e. the successful path, will be reached first, and thus the test will pass before it's had a chance to receive the filtered message. I strongly recommend you take a more robust approach where perhaps you publish 10 messages, 5 of which are filtered, then when a non-filtered channel receives 10, you check that the filtered channel has received 5 and continue.

Reply:
I think the message data here is misleading, the first message is actually the one that will be filtered, as it does not include `headers.number: 26095`. Therefore, if the filter was not working, the subscription should receive this message first and fail.

So, for clarity, I'd suggest we change "no filter on message" to e.g. "no header on message".

Reply:
Ah, yes, you are right... good spot. Thanks for clearing up.

I do think it would be good regardless to have a filtered & non-filtered channel open at the same time to ensure this is possible and we're not somehow mistakenly applying filters to that channel as opposed to that instance of the channel.